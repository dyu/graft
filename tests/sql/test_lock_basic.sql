-- initialize two connections to the same database
.connection 0
.open file:GontkHa6QVLHYnkyk16wUP?vfs=graft
pragma graft_status;

.connection 1
.open file:GontkHa6QVLHYnkyk16wUP?vfs=graft
pragma graft_status;

-- load the sample dataset
.read datasets/simple.sql

-- scenario: lock a table in one connection, then try to lock it in the other
.connection 0
begin immediate;
.connection 1
begin immediate;

-- reset
.connection 0
rollback;
.connection 1
rollback;

-- scenario: verify that upgrading is refused if a read snapshot is outdated
.connection 0
begin;
-- take a read lock
select count(*) from t;

.connection 1
-- update the table, this autocommits
insert into t values(1);

.connection 0
-- try to upgrade the lock via performing a write; this should fail
insert into t values(2);

-- reset
.connection 0
rollback;
.connection 1
rollback;

-- scenario: verify that we can commit a write tx while another tx holds a read lock

-- take a write lock
.connection 0
begin immediate;

-- take a read lock
.connection 1
begin;
select count(*) from t;

-- upgrade our write lock to Pending
.connection 0
insert into t values('committed while read lock is held');

-- try to commit; this should work because the read lock is not blocking
commit;

-- back on the read conn, verify that we still don't see the new row
.connection 1
select * from t;

-- commit the read tx
commit;

-- verify that we now see the new row
select * from t;

-- check metadata
pragma graft_status;
.connection 0
pragma graft_status;
