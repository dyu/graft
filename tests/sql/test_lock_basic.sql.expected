.load libgraft

-- initialize two connections to the same database
.connection 0
.open file:GontkHa6QVLHYnkyk16wUP?vfs=graft
pragma graft_status;
+------------------------+
|      Graft Status      |
+------------------------+
| Graft Status           |
| Current snapshot: None |
| Autosync: false        |
+------------------------+

.connection 1
.open file:GontkHa6QVLHYnkyk16wUP?vfs=graft
pragma graft_status;
+------------------------+
|      Graft Status      |
+------------------------+
| Graft Status           |
| Current snapshot: None |
| Autosync: false        |
+------------------------+

-- load the sample dataset
.read datasets/simple.sql
BEGIN TRANSACTION;
CREATE TABLE t (data text);
INSERT INTO t VALUES('hello world');
INSERT INTO t VALUES('hi bob');
INSERT INTO t VALUES('testing');
COMMIT;


-- scenario: lock a table in one connection, then try to lock it in the other
.connection 0
begin immediate;
.connection 1
Runtime error near line 19: database is locked (5)
begin immediate;

-- reset
.connection 0
rollback;
.connection 1
(1) statement aborts at 1: [rollback;] cannot rollback - no transaction is active
Runtime error near line 25: cannot rollback - no transaction is active
rollback;

-- scenario: verify that upgrading is refused if a read snapshot is outdated
.connection 0
begin;
-- take a read lock
select count(*) from t;
+----------+
| count(*) |
+----------+
| 3        |
+----------+

.connection 1
-- update the table, this autocommits
insert into t values(1);

.connection 0
-- try to upgrade the lock via performing a write; this should fail
Runtime error near line 39: database is locked (5)
insert into t values(2);

-- reset
.connection 0
rollback;
.connection 1
(1) statement aborts at 1: [rollback;] cannot rollback - no transaction is active
Runtime error near line 45: cannot rollback - no transaction is active
rollback;

-- scenario: verify that we can commit a write tx while another tx holds a read lock

-- take a write lock
.connection 0
begin immediate;

-- take a read lock
.connection 1
begin;
select count(*) from t;
+----------+
| count(*) |
+----------+
| 4        |
+----------+

-- upgrade our write lock to Pending
.connection 0
insert into t values('committed while read lock is held');

-- try to commit; this should work because the read lock is not blocking
commit;

-- back on the read conn, verify that we still don't see the new row
.connection 1
select * from t;
+-------------+
|    data     |
+-------------+
| hello world |
| hi bob      |
| testing     |
| 1           |
+-------------+

-- commit the read tx
commit;

-- verify that we now see the new row
select * from t;
+-----------------------------------+
|               data                |
+-----------------------------------+
| hello world                       |
| hi bob                            |
| testing                           |
| 1                                 |
| committed while read lock is held |
+-----------------------------------+

-- check metadata
pragma graft_status;
+---------------------------------+
|          Graft Status           |
+---------------------------------+
| Graft Status                    |
| Current snapshot: Snapshot[3;2] |
| Autosync: false                 |
+---------------------------------+
.connection 0
pragma graft_status;
+---------------------------------+
|          Graft Status           |
+---------------------------------+
| Graft Status                    |
| Current snapshot: Snapshot[3;2] |
| Autosync: false                 |
+---------------------------------+

SQLite Exit Code = 1
